contracts/contracts/token/OUSD.sol

```diff
@@ -234,32 +256,89 @@ contract OUSD is Initializable, InitializableERC20Detailed, Governable {
-            // are removed from the non rebasing tally
-            nonRebasingSupply = nonRebasingSupply.add(_value);
-            // Update rebasingCredits by subtracting the deducted amount
-            _rebasingCredits = _rebasingCredits.sub(creditsDeducted);
-        } else if (!isNonRebasingTo && isNonRebasingFrom) {
-            // Transfer to rebasing account from non-rebasing account
-            // Decreasing non-rebasing credits by the amount that was sent
-            nonRebasingSupply = nonRebasingSupply.sub(_value);
-            // Update rebasingCredits by adding the credited amount
-            _rebasingCredits = _rebasingCredits.add(creditsCredited);
+    }
+
+    function _adjustAccount(address account, int256 balanceChange)
+        internal
+        returns (int256 rebasingCreditsDiff, int256 nonRebasingSupplyDiff)
+    {
+        RebaseOptions state = rebaseState[account];
+        int256 currentBalance = balanceOf(account).toInt256();
+        if (currentBalance + balanceChange < 0) {
+            revert("Transfer amount exceeds balance");
+        }
+        uint256 newBalance = (currentBalance + balanceChange).toUint256();
+
+        if (state == RebaseOptions.YieldDelegationSource) {
+            address target = yieldTo[account];
+            uint256 targetOldBalance = balanceOf(target);
+            uint256 targetNewCredits = _balanceToRebasingCredits(
+                targetOldBalance + newBalance
+            );
+            rebasingCreditsDiff =
+                targetNewCredits.toInt256() -
+                _creditBalances[target].toInt256();
+
+            _creditBalances[account] = newBalance;
+            _creditBalances[target] = targetNewCredits;
+        } else if (state == RebaseOptions.YieldDelegationTarget) {
+            uint256 newCredits = _balanceToRebasingCredits(
+                newBalance + _creditBalances[yieldFrom[account]]
+            );
+            rebasingCreditsDiff =
+                newCredits.toInt256() -
+                _creditBalances[account].toInt256();
+            _creditBalances[account] = newCredits;
+        } else {
+            _autoMigrate(account);
+            if (alternativeCreditsPerToken[account] > 0) {
+                nonRebasingSupplyDiff = balanceChange;
+                alternativeCreditsPerToken[account] = 1e18;
+                _creditBalances[account] = newBalance;
+            } else {
+                uint256 newCredits = _balanceToRebasingCredits(newBalance);
+                rebasingCreditsDiff =
+                    newCredits.toInt256() -
+                    _creditBalances[account].toInt256();
+                _creditBalances[account] = newCredits;
+            }
+        }
+    }
+
+    function _adjustGlobals(
+        int256 rebasingCreditsDiff,
+        int256 nonRebasingSupplyDiff
+    ) internal {
+        if (rebasingCreditsDiff != 0) {
+            if (_rebasingCredits.toInt256() + rebasingCreditsDiff < 0) {
```

Comment:
We should be able to remove these paired underflow checks. We don't need a custom message for these - they should never happen in real life, and safecasts will revert if they do. We don't do duplicate checks for these in the other places where we modified these values.

Reply:
https://github.com/OriginProtocol/origin-dollar/pull/2298/commits/24865d02a9f0e29116f72b71490eed0dbee375df